Add more Pandas-based Checkpointing and Save/Load Functions

ilumsden opened the pull request:
Follow up to https://github.com/hatchet/hatchet/pull/272

This PR adds the following new functions for checkpointing GraphFrames (i.e., saving to/reading from files):

1. `to_pickle` and `from_pickle` ([Pickle Format](https://docs.python.org/3/library/pickle.html#:~:text=%E2%80%9CPickling%E2%80%9D%20is%20the%20process%20whereby,back%20into%20an%20object%20hierarchy.))
2. `to_csv` and `from_csv`
3. `to_excel` and `from_excel`
These functions utilize similar read/write functions from Pandas. In many cases, these Pandas functions require additional dependencies. Those dependencies will not be required in Hatchet. If the dependency for a particular function is not installed, Pandas will raise an `ImportError`.

This PR also adds new `save` and `load` functions to the `GraphFrame` class. These functions can be used to simplify the use of checkpointing. Both of these functions only require one argument: the filename. If the filename contains a recognized extension, that format will be used. Otherwise, the optional `fileformat` parameter can be provided to specify the desired format. If the necessary dependencies are not installed, the `ImportError` raised by Pandas will be caught. In that case, all remaining formats will be attempted. If no supported format succeeds, an `IOError` will be raised.

All the new functions added in this PR accepts keyword arguments (i.e., `**kwargs`). These arguments will be passed to the Pandas function that is eventually invoked to read/write the file. Documentation (i.e., docstrings) will be added that will link to the associated functions' documentation.

Other file formats (e.g., Parquet and Feather) will be added in future PRs.

tgamblin commented:
Travis is failing, but minor note: this is not a "cast".  It's constructing a new list from the view.

ilumsden committed:
Adds Pickle reader/writer

ilumsden committed:
Creates Readers and Writers for CSV and Excel for checkpointing

ilumsden committed:
Gets the CSV and Excel Readers and Writers working by manually picling nodes and converting string representations of lists back to lists

ilumsden committed:
Adds the necessary dependencies for Excel reading and writing to CI

ilumsden committed:
Adds a Python verison guard to the use of ast.literal_eval for pickled nodes because of the lack of bytes type in Python 2

ilumsden committed:
Changes keyword argument parsing for to_hdf and from_hdf to use **kwargs

ilumsden committed:
Changes keyword argument parsing for to_hdf and from_hdf to use **kwargs

Adds all possible optional dependencies for checkpointing to CI

Changes keyword argument parsing for to_hdf and from_hdf to use **kwargs

Adds all possible optional dependencies for checkpointing to CI

ilumsden committed:
Updates CSV, Excel, and Pickle readers and writers to account for changes merged from develop

ilumsden committed:
Removes manual parsing of kwargs in HDF5Reader and HDF5Writer

ilumsden committed:
Merge branch 'develop' of github.com:LLNL/hatchet into pandas_save_load
